src/hotspot/share/opto/parse1.cpp

```diff
@@ -1896,15 +1896,18 @@ void Parse::merge_common(Parse::Block* target, int pnum) {
           ObjID id = as.is_alias(phi);
 
           if (as.contains(id)) {
-            if (pred_as.contains(id)) {
-              // id are in both AS and PRED_AS
+            if (pred_as.is_alias(n) == id) {
+              // merge the same object but have different allocation states.
               if (!as.get_object_state(id)->is_virtual() && pred_as.get_object_state(id)->is_virtual()) {
                 n = ensure_object_materialized(n, pred_as, newin, r, pnum);
               } else if (as.get_object_state(id)->is_virtual() && !pred_as.get_object_state(id)->is_virtual()) {
```

Comment:
I suspect we can never hit this block because `as.get_object_state(id)` will always evaluate to false. Since we pessimistically set all merged objects to escaped, I'm not seeing how any object is ever virtual.

Reply:
This patch only sets an object Escaped when Parse realizes that it's merging a reference that is pointing to single ObjIDs. eg. 
phi(nullptr, obj1), 
phi(obj1, nullptr),
phi(obj1, obj2)

Because the resultant phi doesn't possess 'unique type property', we bail out. 

I think it's still possible that as.get_object_state(id)->is_virtual() is true at this point. let's side track it in JVM-1889

Reply:
Yeah, I think those scenarios you listed should be common enough that we don't see the assertion show up in java base modules.